qcs404.c

```diff
@@ -49,147 +50,190 @@
 #define GCC_CLOCK_FREQ_MEASURE_STATUS	0x74008
 #define GCC_XO_DIV4_CBCR		0x30034
 
-static struct debug_mux gcc;
+struct turing_mux {
+	struct debug_mux mux;
 
-static struct debug_mux gcc = {
-	.phys =	GCC_BASE,
-	.size = GCC_SIZE,
+	unsigned int ahb_reg;
+	unsigned int ahb_mask;
+};
+
+unsigned long measure_turing(const struct measure_clk *clk,
+			     const struct debug_mux *mux);
+
+static struct gcc_mux gcc = {
+	.mux = {
+		.phys =	GCC_BASE,
+		.size = GCC_SIZE,
 
-	.enable_reg = GCC_DEBUG_CLK_CTL,
-	.enable_mask = BIT(16),
+		.measure = measure_gcc,
 
-	.mux_reg = GCC_DEBUG_CLK_CTL,
-	.mux_mask = 0x1ff,
+		.enable_reg = GCC_DEBUG_CLK_CTL,
+		.enable_mask = BIT(16),
 
-	.div_reg = GCC_DEBUG_CLK_CTL,
-	.div_shift = 12,
-	.div_mask = 0xf << 12,
-	.div_val = 4,
+		.mux_reg = GCC_DEBUG_CLK_CTL,
+		.mux_mask = 0x1ff,
+
+		.div_reg = GCC_DEBUG_CLK_CTL,
+		.div_shift = 12,
+		.div_mask = 0xf << 12,
+		.div_val = 4,
+	},
 
 	.xo_div4_reg = GCC_XO_DIV4_CBCR,
 	.debug_ctl_reg = GCC_CLOCK_FREQ_MEASURE_CTL,
 	.debug_status_reg = GCC_CLOCK_FREQ_MEASURE_STATUS,
 };
 
-static struct debug_mux turing = {
-	.phys = 0x800000,
-	.size = 0x30000,
+static struct turing_mux turing = {
+	.mux = {
+		.phys = 0x800000,
+		.size = 0x30000,
 
-	.enable_reg = 0x22008,
-	.enable_mask = BIT(0),
+		.measure = measure_turing,
+		.parent = &gcc.mux,
+		.parent_mux_val = 50,
```

Comment:
Issue with the original code: unfortunate that this isn't hex, while everything else is.

Reply:
qcs404, sdm845 and msm8936 use decimal values instead of hex. I didn't touch that, since it is out-of-scope for this PR.

Reply:
Opened this as a separate issue (#27 )